### DISCUSSION TITLE
Description of the algorithm for Trio before oref0 call : override, temp target and TDD

### DISCUSSION THREAD

**Discussion by avouspierre** [nightscout/Trio#214] [General]
I analysed the different part of code where Trio prepares data to send to oref0 algorithms. It seems important to check the specification is correct before any eventual changes in the code. I also checked in iAPS V4 where a part of code has moved in prepare.js : 

## Override 

Override UX allows to define : 
- override pourcentage : 0 to 200 % 
- ISF option : no guardrail
- CR option : no guardrail
- Max SMB Basal minutes : no guardrail
- Max UAM Basal minutes : no guardrail
- a new BG target : no guardrail
- disable SMB 
- disable SMB durant a period 
- change max IOB (not in Trio) 

Override is used to modify the profile before oref0 call : 
- change the basal value (current value of basal * override %)
- if option ISF, change the ISF (* override %)
- if option CR, change the Carb ratio  (* override %)
- if max SMB, apply the value proposed by override as a new max SMB
- if max UAM, apply the value proposed by override as a new max UAM
- if BG target AND not a temp target, the profile is updated with min and max target bg 
- if disable SMB, microbolus is not allowed
- if disable SMB durant a period, microbolus is not allowed if looping is in the range of the period 
_- if new max IOB, change the value of max IOB in the profile (not in Trio)_

## Temp Target

Temp target allows to define :
- a pourcentage of insulin between 15% and 200 %
- the value of and half basal exercice target (hbt) between 101 to 295 mg/dl
- a duration 
- a date

Temp target is used to modify the profile before oref0 call : 
- the % of insulin modify the target  with the formula in mg/dl : 

$$ \text{low Target BG} = \frac{HBT-100}{\text{pourcentage Of Insulin}}-HBT+200 $$

_Except : ((140-100)/1,2)-140+200 = 116_ 

**Guard : if $\frac{HBT-100}{\text{pourcentage Of Insulin}} * (HBT-100) \lt 0$ ➡️ maxAutosens is used as %** 

- the HBT changes the low target BG with the previous value and is sent to oref0 in the profile.

the date allows to manage temp target in the future (or in the past). 

## TDD 

the algo integrates the TDD as explained here in the [documentation](https://docs.diy-trio.org/en/latest/settings/configuration/preferences/dynamicsettings.html#weighted-average-of-tdd-weight-of-past-24-hours 
).

**TDD is calculated with values modified by override (ISF, basal…)** 

If BG target > 118 and exercice mode activated (highTemptargetRaisesSensitivity or exerciseMode) ➡️ No Dynamic ISF.

if autosens max = min or if max < 1 or min > 1 ➡️ no dynamic ISF.

### Calculation of factor dISF in function of peak insulin : 
| Insulin type | factor insulin in dISF |
|--------|--------|
| rapid-acting | 55 |
| ultra-rapid | 70 |
| custom peak | 120 - custom peak | 
	
### Calculation of weight average of Insulin (TDD) 
$$ \text{Average 2 hours} = \sum \text{Insulin injection TDD from 2 previous hours} / \text{Number of insulin injection} $$

$$ \text{Average 10 days (not 14 !) } = \sum \text{Insulin injection during 10 last days} / \text{Number of insulin injection} $$

$$ TDD = (\text{Weighted of past 2 hours} * \text{average Total TDD from 2 previous 2 hours}) + (1 - \text{Weighted of past 2 hours}) * \text{average 10 days} $$ 

### Algo logarithm  
$$ Ratio = \frac{ISF * \text{adjustement factor} * TDD * \log((\text{Current BG} / \text{factor insulin})+1)}{1800} $$

_Except: 60 * 0,8 * 52 * log(100/55 + 1) / 1800 = 0,62_ 

### Algo Sigmoid 
$$  \Delta{autosens} = \text{autosens max} - \text{autosens min} $$

$$ \Delta{ \text{BG (in mmol/l)}} = (\text{current BG} -  \text{target BG}) * 0,0555 $$

autosens max <> 1 : 

$$ Foffset = \frac{\log{\frac{1}{\text{autosens max} -1} -\frac{\text{autosens min}}{{\text{autosens max} -1}}}}{ \log{e}}  $$

autosens max = 1 : 

$$ Foffset = \frac{\log{\frac{1}{0.01} -\frac{\text{autosens min}}{{0.01}}}}{ \log{e}}  $$

$$ Exponent = \text{deviation BG} * \text{adjustement factor} * TDD + Foffset $$ 

$$ ratio = \frac{\Delta{autosens}}{( 1 + \exp(-Exponent))} + \text{autosens min} $$ 

Check the ratio is in the range of autosens min / max.

- if option ISF, change the ISF (* 1/ratio)
- if option CR, change the Carb ratio  (* 1/ratio)
- if option Basal, change the basal value (current value of basal * ratio)

## Others : 
in iAPS, a dynamic variable is available : disable flat CGM errors : 
- if disable flat CGM and >2 values of BG:
        - if the difference between BG(n) and BG(n-1) is less than 5 mg/l, then change BG(n-1) with + or - 5 mg/l  (in function of the sense) avoided oref0 to consider to be a flat change

**Comment by avouspierre** [comment]
In regard of the documentation, I noticed : 
- SOLVED : _TDD is not calculated with 24 hours and 14 days but 2 hours and 10 days_
- Not sure of the algo with autosens max = 1 for sigmoid algorithm for TDD 
- not found in the doc the temp target description in particular for the change of HBT. 

To check @MikePlante1  @bjornoleh  and others oref0 experts

**Comment by bjornoleh** [comment]
Thanks for the analysis! 

Just a quick note for now:

* if option ISF, change ISF by (* 1/ratio)

* if option CR, change the Carb ratio (* 1/ratio)

**Comment by dnzxy** [comment]
Great summary Pierre, thanks for collecting all that info 👏 

Around profiles and SMB enablement / disablement, iAPS also introduced a hard limit for HIGH glucose readings (it’s hidden from the actual configs of the overrides in the UI and swift code, but present in JS): 

`else if (bg == 400) {
            console.error("Invalid CGM (HIGH). SMBs disabled.");
        return false;
    }`

Found here in `enableSMB` https://github.com/nightscout/trio-oref/blob/b45483738b5f978399f3e91b96eab6c457a614d3/lib/determine-basal/determine-basal.js#L47

Just to add to this - it’s somewhat connected to overrides, but only indirectly. Wanted to mention it here.

**Reply by mountrcg** [reply]
The variables 
- `maxSMBBasalMinutes`
- `maxUAMSMBBasalMinutes`
have no guard rails in oref and therefore not in Trio. 

AAPS has a guardrail implemented at 120min for both. Several hacks exist to circumvent them and allow higher SMB's.
